extensions/amp-viewer-integration/0.1/messaging.js

```diff
@@ -0,0 +1,178 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+import {listen} from '../../../src/event-helper';
+import {dev} from '../../../src/log';
+
+
+/**
+ * @fileoverview This is the communication protocol between AMP and the viewer.
+ * It is used in the amp-viewer-integration.js file.
+ */
+export class Messaging {
+  /**
+   * Messaging protocol between viewer and viewer client.
+   * @param {!Window} target
+   * @param {string} targetOrigin
+   * @param {function(string, *, boolean):(!Promise<*>|undefined)}
+   *    requestProcessor
+   */
+  constructor(target, targetOrigin, requestProcessor) {
+    this.sentinel_ = '__AMP__';
+    this.requestSentinel_ = this.sentinel_ + 'REQUEST';
+    this.responseSentinel_ = this.sentinel_ + 'RESPONSE';
+
+    this.requestIdCounter_ = 0;
+    this.waitingForResponse_ = {};
+
+    /** @const @private {!Window} */
+    this.target_ = target;
+    /** @const @private {string} */
+    this.targetOrigin_ = targetOrigin;
+    /** @const @private {function(string, *, boolean):(!Promise<*>|undefined)} */
+    this.requestProcessor_ = requestProcessor;
+
+    dev().assert(this.targetOrigin_, 'Target origin must be specified!');
+
+    listen(this.target_, 'message', (this.onMessage_.bind(this)));
+  }
+
+
+  /**
+   * @param {Event|null} event
+   * @private
+   */
+  onMessage_(event) {
+    if (!event || event.source != this.target_ ||
+      event.origin != this.targetOrigin_) {
+      return;
+    }
+    const message = event.data;
+    if (message.sentinel == this.requestSentinel_) {
+      this.onRequest_(message);
+    } else if (message.sentinel == this.responseSentinel_) {
```

Comment:
Are both "responses" and "requests" messages that are being sent from the parent to this window?